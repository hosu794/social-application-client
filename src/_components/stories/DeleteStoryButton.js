import React from "react";

function DeleteStoryButton({ deleteHandler }) {
  return (
    <button onClick={deleteHandler} className="button is-danger">
      Delete
    </button>
  );
}

export default DeleteStoryButton;
