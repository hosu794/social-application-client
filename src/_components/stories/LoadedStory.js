import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { storyActions, userActions } from "../../_actions";
import { useDispatch, useSelector } from "react-redux";
import { IUser, IStory } from "../../_types";

import CastStoreButton from "./CastStoreButton";

import UnCastStoreButton from "./UnCastStoreButton";

import DeleteStoryButton from "./DeleteStoryButton";
import { Link } from "react-router-dom";

function LoadedStory(props) {
  const dispatch = useDispatch();

  const editStoryUrl = `/stories/edit/${props.id}`;

  const isUserLovedStory = useSelector((state) => state.user.isUserLovedStory);
  const isLogged = useSelector((state) => state.authentication.loggedIn);
  const user = useSelector((state) => state.user.user);
  const isUserExist = user ? user.id : null;
  const isUserIdentificationIsEqualStoryCreator = isUserExist === props.userId;
  const currentStory = useSelector((state) => state.stories.currentStory);

  function deleteHandler() {
    dispatch(storyActions.deleteStory(props.id));
  }

  const unCastStore = async () => {
    await dispatch(storyActions.unCastLove({ storyId: props.id }));
  };

  const castStore = async () => {
    await dispatch(storyActions.castLove({ storyId: props.id }));
  };

  const checkLoveAvailabilityIfIsLogged = async () => {
    if (isLogged) {
      await dispatch(
        userActions.checkLoveAvailability({
          storyId: currentStory.id,
          userId: isUserExist,
        })
      );
    }
  };

  useEffect(() => {
    checkLoveAvailabilityIfIsLogged();
  }, [props.loves]);

  return (
    <section className="section">
      <div className="container">
        <h1 className="title">{props.title}</h1>
        <h2 className="subtitle">{props.description}</h2>
        <p dangerouslySetInnerHTML={{ __html: props.body }}></p>
        <nav
          className="level"
          style={{
            marginTop: "1em",
          }}
        >
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Author</p>
              <p className="title">{props.creator.username}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Topic</p>
              <p className="title">{props.topic.title}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Loves</p>
              <p className="title">{props.loves}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            {isLogged ? (
              <div>
                {isUserLovedStory ? (
                  <CastStoreButton id={props.id} castStore={castStore} />
                ) : (
                  <UnCastStoreButton id={props.id} unCastStore={unCastStore} />
                )}
              </div>
            ) : null}
          </div>
          {isUserIdentificationIsEqualStoryCreator ? (
            <div class="level-item has-text-centered">
              <div>
                <DeleteStoryButton deleteHandler={deleteHandler} />
              </div>
            </div>
          ) : null}
          {isUserIdentificationIsEqualStoryCreator ? (
            <div class="level-item has-text-centered">
              <div>
                <Link to={editStoryUrl}>
                  <button className="button is-info">Edit</button>
                </Link>
              </div>
            </div>
          ) : null}
        </nav>
      </div>
    </section>
  );
}

LoadedStory.propTypes = {
  isUserLovedStory: PropTypes.bool,
  user: PropTypes.instanceOf(IUser),
  props: PropTypes.shape({
    title: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    body: PropTypes.string.isRequired,
    loves: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    creator: PropTypes.shape({
      username: PropTypes.string.isRequired,
    }),
  }),
};

export default LoadedStory;
