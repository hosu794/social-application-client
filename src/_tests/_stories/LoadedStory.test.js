import React from "react";

import configureStore from "redux-mock-store";
import { useDispatch } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import LoadedStory from "../../_components/stories/LoadedStory";
import { mount } from "enzyme";

import { Provider } from "react-redux";

import DeleteStoryButton from "../../_components/stories/DeleteStoryButton";

const mockStore = configureStore([]);

let store = mockStore({
  user: {
    isUverLovedStory: true,
    user: {
      id: 12,
      name: "Joe Doe",
      username: "exampleusername",
      downloadAvatar: "dsdasdasdasdasdadas.png",
    },
  },
  authentication: {
    loggenIn: true,
  },
  stories: {
    currentStory: {
      id: 1,
      title: "Story Title",
      description: "Story description",
      totalLoves: 12,
      body: "<p>Simple Story</p>",
      createdBy: {
        id: 12,
        name: "Joe Doe",
        username: "exampleusername",
        downloadAvatar: "dsdasdasdasdasdadas.png",
      },
      topic: {
        id: 12,
        title: "Topic Example Title",
        description: "Description Example description",
        createdAt: "Dummy date",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    },
  },
});

const story = {
  id: 1,
  title: "Story Title",
  description: "Story description",
  totalLoves: 12,
  body: "<p>Simple Story</p>",
  createdBy: {
    id: 12,
    name: "Joe Doe",
    username: "exampleusername",
    downloadAvatar: "dsdasdasdasdasdadas.png",
  },
  topic: {
    id: 12,
    title: "Topic Example Title",
    description: "Description Example description",
    createdAt: "Dummy date",
    createdBy: {
      id: 12,
      name: "Joe Doe",
      username: "exampleusername",
      downloadAvatar: "dsdasdasdasdasdadas.png",
    },
  },
  authentication: {
    loggedIn: true,
  },
};

describe("Test for the LodaedStory Component ", () => {
  let TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <LoadedStory
          title={story.title}
          description={story.description}
          key={story.id}
          creator={story.createdBy}
          loves={story.totalLoves}
          topic={story.topic}
          body={story.body}
          id={story.id}
          user={story.createdBy}
          userId={story.createdBy.id}
        />
      </Provider>
    </MemoryRouter>
  );

  let wrapper = mount(<TestingComponent />);

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(LoadedStory).toBeDefined();
  });

  test("should render correctly story title", () => {
    let element = wrapper.find(".title").first().text();
    expect(element).toEqual(story.title);
  });

  test("should render correctly story description", () => {
    let element = wrapper.find(".subtitle").first().text();
    expect(element).toEqual(story.description);
  });

  test("should render correctly body ", () => {
    let element = wrapper.find("p").first().text();

    expect(element).toEqual("Simple Story");
  });

  test("should render correctly author ", () => {
    let element = wrapper.find(".title").at(1).text();

    expect(element).toEqual(story.createdBy.username);
  });

  test("should render correctly topic title", () => {
    let element = wrapper.find(".title").at(2).text();

    expect(element).toEqual(story.topic.title);
  });

  test("should render correctly count of loves", () => {
    let element = wrapper.find(".title").at(3).text();

    expect(element).toEqual("12");
  });

  test("should return DeleteButton Component", () => {
    let isExist = wrapper.exists(DeleteStoryButton);

    expect(isExist).toEqual(true);
  });
});
