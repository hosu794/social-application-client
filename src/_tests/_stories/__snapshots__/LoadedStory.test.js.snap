// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the LodaedStory Component  should render component correctly 1`] = `ReactWrapper {}`;
